subsys/bluetooth/audio/has_client.c

```diff
@@ -0,0 +1,770 @@
+	{
+		return -EINVAL;
+	}
+
+	if (id == BT_HAS_PRESET_INDEX_NONE) {
+		return -EINVAL;
+	}
+
+	hdr = (void *)buf;
+
+	/* Request synchronization of Binaural Set if supported */
+	if (BT_HAS_PRESET_SYNC_MASK(has->features)) {
+		hdr->op = BT_HAS_OP_SET_ACTIVE_PRESET_SYNC;
+	} else {
+		hdr->op = BT_HAS_OP_SET_ACTIVE_PRESET;
+	}
+
+	req = (void *)hdr->data;
+	req->idx = id;
+
+	return control_point_tx(client, client_preset_active_set_cb, buf, sizeof(buf));
+}
+
+int bt_has_client_preset_active_set_next(struct bt_has *has)
+{
+	struct bt_has_client *client = BT_HAS_CLIENT(has);
+	struct bt_has_cp_hdr hdr;
+
+	/* Request synchronization of Binaural Set if supported */
+	if (BT_HAS_PRESET_SYNC_MASK(has->features)) {
+		hdr.op = BT_HAS_OP_SET_NEXT_PRESET_SYNC;
+	} else {
+		hdr.op = BT_HAS_OP_SET_NEXT_PRESET;
+	}
+
+	return control_point_tx(client, client_preset_active_set_cb, &hdr, sizeof(hdr));
+}
+
+int bt_has_client_preset_active_set_prev(struct bt_has *has)
+{
+	struct bt_has_client *client = BT_HAS_CLIENT(has);
+	struct bt_has_cp_hdr hdr;
+
+	/* Request synchronization of Binaural Set if supported */
+	if (BT_HAS_PRESET_SYNC_MASK(has->features)) {
+		hdr.op = BT_HAS_OP_SET_PREV_PRESET_SYNC;
+	} else {
+		hdr.op = BT_HAS_OP_SET_PREV_PRESET;
+	}
+
+	return control_point_tx(client, client_preset_active_set_cb, &hdr, sizeof(hdr));
+}
+
+static int bt_has_read_preset_req(struct bt_has_client *client, uint8_t start_idx, uint8_t count)
+{
+	struct bt_has_cp_hdr *hdr;
+	struct bt_has_cp_read_preset_req *req;
+	uint8_t buf[sizeof(*hdr) + sizeof(*req)];
+
+	hdr = (void *)buf;
+	hdr->op = BT_HAS_OP_READ_PRESET_REQ;
+	req = (void *)hdr->data;
+	req->start_id = start_idx;
+	req->num_presets = count;
+
+	return control_point_tx(client, client_preset_read_cb, buf, sizeof(buf));
+}
+
+int bt_has_client_preset_read(struct bt_has *has, uint8_t idx)
+{
+	struct bt_has_client *client = BT_HAS_CLIENT(has);
+	int err;
+
+	if (client->read_preset_idx) {
+		return -EBUSY;
+	}
+
+	err = bt_has_read_preset_req(client, idx, 1u);
+	if (err == 0) {
+		client->read_preset_idx = idx;
+	}
+
+	return err;
+}
+
+int bt_has_client_preset_read_multiple(struct bt_has *has, uint8_t start_idx, uint8_t count)
+{
+	return bt_has_read_preset_req(BT_HAS_CLIENT(has), start_idx, count);
+}
+
+int bt_has_client_preset_name_set(struct bt_has *has, uint8_t id, const char *name)
+{
+	struct bt_has_client *client = BT_HAS_CLIENT(has);
+	struct bt_has_cp_hdr *hdr;
+	struct bt_has_cp_write_preset_name_req *req;
+	uint8_t buf[sizeof(*hdr) + sizeof(*req) + strlen(name)];
+
+	/* FIXME */
+
+	if (strlen(name) < BT_HAS_PRESET_NAME_MIN || strlen(name) > BT_HAS_PRESET_NAME_MAX) {
+		return -EINVAL;
+	}
+
+	hdr = (void *)buf;
+	hdr->op = BT_HAS_OP_WRITE_PRESET_NAME;
+	req = (void *)hdr->data;
+	req->idx = id;
+	strncpy(req->name, name, strlen(name));
+
+	return control_point_tx(client, client_preset_active_set_cb, buf, sizeof(buf));
+}
+
+int bt_has_discover(struct bt_conn *conn)
```

Comment:
It would have helped me if the function had a few comment lines on the top describing the general approach:
- discover first characteristic, subscribe to it, read it, discover next characteristic, ...

Some of the other audio services has this (e.g. VCS, see below).  And as this service follows another pattern, an explanation here might also be in place.

(From vcs_client.c:)
	/*
	 * This will initiate a discover procedure. The procedure will do the
	 * following sequence:
	 * 1) Primary discover for the VCS
	 * 2) Characteristic discover of the VCS
	 * 3) Discover services included in VCS (VOCS and AICS)
	 * 4) For each included service found; discovery of the characteristics
	 * 5) When everything above have been discovered, the callback is called
	 */